CoseSign1.Certificates/Local/Validators/X509ChainTrustValidator.cs

```diff
@@ -129,32 +134,44 @@ public class X509ChainTrustValidator(
             }
         }
 
-        // Chain build failed, but if the only failure is Untrusted Root we may still pass.
-        if (ChainBuilder.ChainStatus.All(st => (st.Status &~ (X509ChainStatusFlags.UntrustedRoot | X509ChainStatusFlags.NoError)) == 0)) // use &~ to mask out UntrustedRoot and NoError
+        // If we're here, chain build failed.
+        // This is the result of building the certificate chain.
+        CoseSign1ValidationResult baseResult = new (GetType(), false,
+            $"[{string.Join("][", ChainBuilder.ChainStatus.Select(cs => cs.StatusInformation).ToArray())}]",
+            ChainBuilder.ChainStatus.Cast<object>().ToList());
+
+        // Ignore failures from untrusted roots or expired certificates if the user tells us to.
+        X509ChainStatusFlags flagsToIgnore = X509ChainStatusFlags.NoError;
+        flagsToIgnore |= AllowUntrusted ? X509ChainStatusFlags.UntrustedRoot : 0;
+
+        // If we have a valid user-supplied root, consider it trusted. (Not supported by .NET Standard 2.0 so we have to do it ourselves.)
+        string chainRootThumb = ChainBuilder.ChainElements.FirstOrDefault(element => element.Subject.Equals(element.Issuer))?.Thumbprint ?? string.Empty;
+        bool trustUserRoot = hasRoots && TrustUserRoots && !string.IsNullOrEmpty(chainRootThumb) && Roots!.Any(r => r.Thumbprint == chainRootThumb);
+        flagsToIgnore |= trustUserRoot ? X509ChainStatusFlags.UntrustedRoot : 0;
+
+        // If allowOutdated is set and none of the outdated certificates in the chain have a lifetime EKU, ignore NotTimeValid.
+        List<X509Certificate2>? outdatedCerts = ChainBuilder.ChainElements?.Where(c => c.NotAfter < DateTime.Now).ToList();
+        bool expired = outdatedCerts?.Any(c => c.Extensions.OfType<X509Extension>().Any(e => e.Oid?.Value == LifetimeEkuOidValue)) ?? false;

```

Comment:
do we have any tests where allowOutdated is true, the cert is expired, and the certificate was created with the lifetime eku?

Reply:
No we don't. I'll add that.

Reply:
Done. And it caught a bug, so I'm glad I made the test.